common/http-api-client/src/dns.rs

```diff
@@ -0,0 +1,186 @@
+//! supporting DoH and DoT.
+//!
+//! This resolver implements a fallback mechanism where, should the DNS-over-TLS resolution fail, a
+//! followup resolution will be done using the hosts configured default (e.g. `/etc/resolve.conf` on
+//! linux).
+//!
+//! Requires the `dns-over-https-rustls`, `webpki-roots` feature for the
+//! `hickory-resolver` crate
+use crate::ClientBuilder;
+
+use std::fmt;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use hickory_resolver::lookup_ip::LookupIp;
+use hickory_resolver::{
+    config::{LookupIpStrategy, NameServerConfigGroup, ResolverConfig, ResolverOpts},
+    error::ResolveError,
+    lookup_ip::LookupIpIntoIter,
+    TokioAsyncResolver,
+};
+use once_cell::sync::OnceCell;
+use reqwest::dns::{Addrs, Name, Resolve, Resolving};
+use tracing::warn;
+
+impl ClientBuilder {
+    pub fn dns_resolver<R: Resolve + 'static>(mut self, resolver: Arc<R>) -> Self {
+        self.reqwest_client_builder = self.reqwest_client_builder.dns_resolver(resolver);
+        self
+    }
+}
+
+struct SocketAddrs {
+    iter: LookupIpIntoIter,
+}
+
+#[derive(Debug)]
+struct HickoryDnsSystemConfError(ResolveError);
+
+/// Wrapper around an `AsyncResolver`, which implements the `Resolve` trait.
+#[derive(Debug, Default, Clone)]
+pub struct HickoryDnsResolver {
+    /// Since we might not have been called in the context of a
+    /// Tokio Runtime in initialization, so we must delay the actual
+    /// construction of the resolver.
+    state: Arc<OnceCell<TokioAsyncResolver>>,
+    fallback: Arc<OnceCell<TokioAsyncResolver>>,
+}
+
+impl Resolve for HickoryDnsResolver {
+    fn resolve(&self, name: Name) -> Resolving {
+        let resolver = self.state.clone();
+        let fallback = self.fallback.clone();
+        Box::pin(async move {
+            let resolver = resolver.get_or_try_init(new_resolver)?;
+
+            // try the primary DNS resolver that we set up (DoH or DoT or whatever)
+            let lookup = match resolver.lookup_ip(name.as_str()).await {
+                Ok(res) => res,
+                Err(e) => {
+                    // on failure use the fall back system configured DNS resolver
+                    warn!("primary DNS failed w/ error {e}: using system fallback");
+                    let resolver = fallback.get_or_try_init(new_resolver_system)?;
+                    resolver.lookup_ip(name.as_str()).await?
+                }
+            };
+
+            let addrs: Addrs = Box::new(SocketAddrs {
+                iter: lookup.into_iter(),
+            });
+            Ok(addrs)
+        })
+    }
+}
+
+impl Iterator for SocketAddrs {
+    type Item = SocketAddr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|ip_addr| SocketAddr::new(ip_addr, 0))
+    }
+}
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+impl HickoryDnsResolver {
+    pub async fn resolve_str(&self, name: &str) -> Result<LookupIp, BoxError> {
```

Comment:
Question, what about just returning the concrete error type?

Reply:
Done